api4/system.go

```diff
@@ -545,3 +548,70 @@ func getServerBusyExpires(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 	w.Write([]byte(c.App.Srv().Busy.ToJson()))
 }
+
+func upgradeToEnterprise(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(*c.App.Session(), model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	auditRec := c.MakeAuditRecord("upgradeToEnterprise", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+
+	percentage, _ := c.App.Srv().UpgradeToE0Status()
+	if model.BuildEnterpriseReady == "true" {
+		c.Err = model.NewAppError("upgradeToEnterprise", "api.upgrade_to_enterprise.already-enterprise.app_error", nil, "", http.StatusTooManyRequests)
+		return
+	}
+	if percentage > 0 {
+		c.Err = model.NewAppError("upgradeToEnterprise", "api.upgrade_to_enterprise.app_error", nil, "", http.StatusTooManyRequests)
+		return
+	}
+	if percentage == 100 {
+		c.Err = model.NewAppError("upgradeToEnterprise", "api.upgrade_to_enterprise.already-done.app_error", nil, "", http.StatusTooManyRequests)
+		return
+	}
+
+	c.App.Srv().Go(func() {
+		c.App.Srv().UpgradeToE0()
+	})
+
+	auditRec.Success()
+	w.WriteHeader(http.StatusAccepted)
+	ReturnStatusOK(w)
+}
+
+func upgradeToEnterpriseStatus(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(*c.App.Session(), model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	percentage, err := c.App.Srv().UpgradeToE0Status()
+	var s map[string]interface{}
+	if err != nil {
+		appErr := model.NewAppError("upgradeToEnterpriseStatus", "api.upgrade_to_enterprise_status.app_error", nil, err.Error(), http.StatusTooManyRequests)
+		s = map[string]interface{}{"percentage": 0, "error": appErr.Error()}
```

Comment:
don't know if it would be useful or not to capture/log/surface that there was an error at a certain percentage, maybe if we can correlate with the file size?

Reply:
I can add that information inside the error when makes sense. For example, only as part of the "downloading error". (See the commit that I'm about to submit).

Reply:
Done